docs/unit_tests.rst

```diff
@@ -0,0 +1,126 @@
+Consequently, unit tests should be clear and concise.
+
+The confidence one has in a piece of code is often related to the number of code paths explored in it.
+This is often approximated by "code coverage."
+That is, one can run the unit test with a coverage tool to see which code paths were exercised by the unit test.
+Code with higher coverage tends to have fewer bugs because the tester has often considered various corner-cases.
+Consequently, unit tests should aim for high code coverage.
+
+Unit tests should be executed frequently to provide developers with instant feedback.
+This applies to the feature under development and the system as a whole.
+That is, developers should frequently execute all of the unit tests to make sure they haven't broken functionality elsewhere in the system.
+The more frequently the tests are run, the smaller the increment of development and the easier it is to identify a breaking change.
+Thus, unit tests should execute quickly.
+
+Code that is difficult to test will most likely be difficult to use.
+Code that is difficult to use correctly will lead to bugs in code that use it.
+Consequently, unit tests are vital to the design of useful software as developing a unit test provides feedback on the design of the code under test.
+Often, when developing a unit test, one will find parts of the design that can be improved.
+
+Unit tests should promote and not inhibit development.
+A robust set of unit tests allows a developer to aggressively refactor since the correctness of the system can be checked after the refactoring.
+However, unit tests do produce drag on development since they must be maintained as the code evolves.
+Thus, it is important that the unit test code be properly maintained so that they are an asset and not a liability.
+
+Some of the goals mentioned above are in conflict.
+Adding code to increase coverage may make the tests less maintainable, slower, and more difficult to understand.
+The following metrics can be generated to measure the utility of the unit tests:
+
+* Code coverage
+* Test compilation time
+* Test execution time
+* Test code size vs. code size
+* Defect rate vs. code coverage (Are bugs appearing is code that is not tested as well?)
+
+Unit Test Organization
+======================
+
+The most basic unit when testing is the *test case*.
+A test case typically has four phases.
+
+1. Setup - The system is initialized to a known state.
+2. Exercise - The code under test is invoked.
+3. Check - The resulting state of the system and outputs are checked.
+4. Teardown - Any resources allocated in the test are deallocated.
+
+Test cases are grouped into a *test suite*.
+
+Test suites are organized into *a test plan*.
+
+We adopt file boundaries for organizing the unit tests for OpenDDS.
+That is, the unit tests for a file group `dds/DCPS/SomeFile.(h|cpp)` will be located in `unit-tests/dds/DCPS/SomeFile.cpp`.
```

Comment:
Shouldn't it be under `tests/unit-tests`? I think the root directory is already too crowded.

Reply:
Sure.  If that's the case, we should have
* `tests/unit-tests`
* `tests/performance-tests`
*  etc.

Reply:
Yeah, if we wanted to be consistent. That would mean moving bench again though after just moving it.